core/src/main/java/io/parsingdata/metal/expression/value/Fold.java

```diff
@@ -57,42 +57,47 @@ public Fold(final ValueExpression values, final BinaryOperator<ValueExpression>
     }
 
     @Override
-    public ImmutableList<Optional<Value>> eval(final ParseState parseState, final Encoding encoding) {
-        final ImmutableList<Optional<Value>> initialList = initial != null ? initial.eval(parseState, encoding) : new ImmutableList<>();
-        if (initialList.size > 1) {
-            return new ImmutableList<>();
+    public ImmutableList<Value> eval(final ParseState parseState, final Encoding encoding) {
+        final ImmutableList<Value> initialList = initial != null ? initial.eval(parseState, encoding) : new ImmutableList<>();
+        if (initialList.size > 1 || initialList.head == NOT_A_VALUE) {
+            return ImmutableList.create(NOT_A_VALUE);
```

Comment:
There are other cases where we throw an `Exception` in this case. Shouldn't we do so here? (Probably a separate issue.)

Reply:
Or should we not throw exceptions elsewhere? Returning `NOT_A_VALUE` here seems fine to me, at first though anyway.

Reply:
We need to evaluate where we throw an exception and where we return `NOT_A_VALUE`. @ccreeten: I tend to agree that the latter will usually be entirely acceptable.